python/paddle/fluid/backward.py

```diff
@@ -1101,67 +1102,119 @@ def append_backward(loss,
+        program, current_block_idx)
+
+    # block_id: fwd_op_num
+    block_fwd_op_num_dict = {}
+    for idx in son_parent_block_idx_dict:
+        block_fwd_op_num_dict[idx] = program.block(idx).desc.op_size()
+
     grad_to_var = dict()
 
     op_desc = _create_loss_op_desc_(loss)
-    root_block.desc.append_op().copy_from(op_desc)
+    target_grad_block.desc.append_op().copy_from(op_desc)
 
-    block_no_grad_set = set(map(_strip_grad_suffix_, no_grad_dict[0]))
-    op_path = _find_op_path_(root_block, [loss], [], block_no_grad_set)
-    no_grad_vars = _find_no_grad_vars(root_block, op_path, [loss],
-                                      block_no_grad_set)
-    block_no_grad_set.update(no_grad_vars)
-    no_grad_dict[0].update(list(map(_append_grad_suffix_, block_no_grad_set)))
+    current_block_no_grad_set = set(
+        map(_strip_grad_suffix_, no_grad_dict[current_block_idx]))
 
-    input_grad_names_set = None
-    # For double backward, input_grad_names is used for filter
-    # some non-used gradients op.
-    if program._appending_grad_times > 1:
-        input_grad_names_set = set([_append_grad_suffix_(loss.name)])
-
-
-    if checkpoints != None and \
-       isinstance(checkpoints, list) and \
-       len(checkpoints) > 0:
-        program_stat, checkpoint_names, \
-        vars_should_be_hold, \
-        recompute_segments = \
-                        _append_backward_ops_with_checkpoints_(
-                            root_block,
-                            op_path,
-                            root_block,
-                            no_grad_dict,
-                            grad_to_var,
-                            checkpoints)
-    else:
-        _append_backward_ops_(
-            root_block,
-            op_path,
-            root_block,
-            no_grad_dict,
-            grad_to_var,
-            callbacks,
-            input_grad_names_set=input_grad_names_set)
+    op_path_dict = {}  # block idx : op_path
+    no_grad_vars_dict = {}
 
-    # Because calc_gradient may be called multiple times,
-    # we need rename the internal gradient variables so that they have
-    # different names.
-    _rename_grad_(root_block, fwd_op_num, grad_to_var, {})
+    for block_idx in son_parent_block_idx_dict:
+        block = program.block(block_idx)
+
+        block_no_grad_set = set(
+            map(_strip_grad_suffix_, no_grad_dict[block_idx]))
+        op_path_dict[block_idx] = _find_op_path_(block, [loss], [],
+                                                 block_no_grad_set)
+
+        no_grad_vars_dict[block_idx] = _find_no_grad_vars(
+            block, op_path_dict[block_idx], [loss], block_no_grad_set)
+
+        block_no_grad_set.update(no_grad_vars_dict[block_idx])
+        no_grad_dict[block_idx].update(
+            list(map(_append_grad_suffix_, block_no_grad_set)))
+
+        input_grad_names_set = None
+        # For double backward, input_grad_names is used for filter
```

Comment:
grammar issue:

"used for filter" -> "used for filtering"
"some non-used gradients op" -> "some non-used gradients op(s)"